### PR TITLE
Enable rt721 support for PTL-RVP

### PR DESCRIPTION
**Pull request by naveen-manohar**
Patch adds match table for rt721 multiple function codec on link3.

BUG=b:368495490
Test=WCL-SDV PO audio tests, verified on PTL-RVP board with RT721 Eval

### CODE REVIEW

File: sound/soc/sdw_utils/soc_sdw_rt721_sdca.c [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#include <sound/soc.h>
+#include <sound/soc-acpi.h>
+#include <sound/soc-dapm.h>
+#include <sound/soc_sdw_utils.h>
+
+static const struct snd_soc_dapm_route rt721_spk_map[] = {
+	{ "Speaker", NULL, "rt721 SPK" },
+};
+
+int asoc_sdw_rt721_spk_rtd_init(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai)
```
Comment (bardliao) [inline sound/soc/sdw_utils/soc_sdw_rt721_sdca.c]: Can we merge those Realtek multi-function codecs spk_rtd_init? Like what we did in soc_sdw_rt_amp.c. Currently, we have soc_sdw_rt712_sdca.c, soc_sdw_rt722_sdca.c, and soc_sdw_rt721_sdca.c, and I am sure there will be more coming.

Comment (naveen-manohar) [reply sound/soc/sdw_utils/soc_sdw_rt721_sdca.c]: Agree, shall respin patch..

File: sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#include <linux/errno.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#include <sound/control.h>
+#include <sound/soc.h>
+#include <sound/soc-acpi.h>
+#include <sound/soc-dapm.h>
+#include <sound/soc_sdw_utils.h>
+
+#define CODEC_NAME_SIZE	7
```
Comment (bardliao) [inline sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c]: Should it be 6? The codec name will be "rt712" , "rt722", etc.

File: sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+int asoc_sdw_rt_mf_sdca_spk_rtd_init(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai)
+{
+	struct snd_soc_card *card = rtd->card;
+	const struct snd_soc_dapm_route *rt_mf_sdca_map;
+	char codec_name[CODEC_NAME_SIZE];
+	struct snd_soc_dai *codec_dai;
+	int ret;
+	int i;
+
+	rt_mf_sdca_map = get_codec_name_and_route(dai, codec_name);
```
Comment (bardliao) [inline sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c]: `get_codec_name_and_route` could return NULL. So, we need to test `if (!rt_mf_sdca_map)`.

File: sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+					  card->components, codec_name);
+	if (!card->components)
+		return -ENOMEM;
+
+	for_each_rtd_codec_dais(rtd, i, codec_dai) {
+		if (strstr(codec_dai->component->name_prefix, "-1"))
+			ret = snd_soc_dapm_add_routes(&card->dapm, rt_mf_sdca_map, 2);
+		else if (strstr(codec_dai->component->name_prefix, "-2"))
+			ret = snd_soc_dapm_add_routes(&card->dapm, rt_mf_sdca_map + 2, 2);
+	}
```
Comment (bardliao) [inline sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c]: I don't think there will be "-1" or "-2" for multi function speakers. Just add `rt_mf_sdca_map` with `ARRAY_SIZE(rt_mf_sdca_map)`

Comment (naveen-manohar) [reply sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c]: respined with a different logic, verified RT722 & RT721 on PTL
please check this

File: sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+		return -ENOMEM;
+
+	/* Add routes */
+	ret = snd_soc_dapm_add_routes(&card->dapm, route_map->route_map, route_map->route_size);
+	if (ret) {
+		dev_err(rtd->dev, "failed to add rt sdca spk map: %d\n", ret);
+		return ret;
+	}
+
+	return 0;
```
Comment (bardliao) [inline sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c]: You could
```c
	if (ret)
		dev_err(rtd->dev, "failed to add rt sdca spk map: %d\n", ret);

	return ret;
```

Comment (naveen-manohar) [reply sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c]: sure, done..

**Review by bardliao — DISMISSED** [review]

File: sound/soc/intel/common/soc-acpi-intel-ptl-match.c [hunk 6]
```diff
@@ -69,6 +69,24 @@ static const struct snd_soc_acpi_adr_device rt711_sdca_0_adr[] = {
...
+	}
+};
+
+static const struct snd_soc_acpi_link_adr ptl_rt721_l3[] = {
+	{
+		.mask = BIT(3),
+		.num_adr = ARRAY_SIZE(rt721_3_single_adr),
+		.adr_d = rt721_3_single_adr,
+	},
+	{}
```
Comment (ujfalusi) [inline sound/soc/intel/common/soc-acpi-intel-ptl-match.c]: nitpick: Would be nice to be consistent within a file and add a comma after the sentinel

**Review by ujfalusi — DISMISSED** [review]

File: sound/soc/intel/common/soc-acpi-intel-ptl-match.c (hunk 6 above)
Comment (naveen-manohar) [reply sound/soc/intel/common/soc-acpi-intel-ptl-match.c]: thank you,
done

**Review by ujfalusi — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**